src/core/ir/discriminator.rs

```diff
@@ -238,7 +244,7 @@ impl Discriminator {
 
     pub fn resolve_type(&self, value: &Value) -> Result<&str> {
         let Value::Object(obj) = value else {
-            bail!("Value expected to be object");
+            return Ok(&self.name);
```

Comment:
I think it could hide some errors when we are trying to resolve object type but getting something else. Null is a special case though, what about return `Option<&str>` here and for null use `None`?